torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -441,57 +428,55 @@ def _pre_forward(
         if unshard_fn is not None:
-            unshard_fn()
+            unshard_fn(state, handle)
         # Register post-backward hooks to reshard the parameters and reduce-scatter
         # their gradients. They must be re-registered every forward pass in case
         # the `grad_fn` is mutated.
-        _register_post_backward_hooks(state, handles)
+        _register_post_backward_hook(state, handle)
         # We have to reallocate the _cpu_grad if optimizer overlap
         # set the grad to None in the backward pass.
-        for handle in handles:
-            if handle._offload_params and handle.flat_param._cpu_grad is None:
-                handle.flat_param._cpu_grad = torch.zeros_like(
-                    handle.flat_param._local_shard, device=torch.device("cpu")
-                ).pin_memory()
-
-        should_cast_forward_inputs = len(state._handles) > 0 and all(
-            not handle._force_full_precision for handle in state._handles
+        if handle and handle._offload_params and handle.flat_param._cpu_grad is None:
+            handle.flat_param._cpu_grad = torch.zeros_like(
+                handle.flat_param._local_shard, device=torch.device("cpu")
+            ).pin_memory()
+
+        should_cast_forward_inputs = (
+            state._handle and not state._handle._force_full_precision
         )
 
         if should_cast_forward_inputs and state.mixed_precision.cast_forward_inputs:
             # Recursively convert args and kwargs to specified precision.
             input_dtype: Optional[torch.dtype] = state.mixed_precision.param_dtype
             args, kwargs = _cast_forward_inputs(input_dtype, *args, **kwargs)
-        _register_post_backward_reshard_only_hooks(state, handles, args, kwargs)
+        _register_post_backward_reshard_only_hook(state, handle, args, kwargs)
         return args, kwargs
 
 
 @no_type_check
 def _pre_forward_unshard(
     state: _FSDPState,
-    handles: List[FlatParamHandle],
+    handle: FlatParamHandle,
 ) -> None:
     """Unshards parameters in the pre-forward."""
-    if not handles:
+    if not handle:
         return
-    handles_key = tuple(handles)
     # If the handles have been prefetched, then there is no need to call
     # `_unshard()` again
-    if not state._handles_prefetched.get(handles_key, False):
-        _unshard(state, handles, state._unshard_stream, state._pre_unshard_stream)
-    state._needs_pre_forward_unshard[handles_key] = False
+    if not state._handles_prefetched.get(handle, False):
+        _unshard(state, handle, state._unshard_stream, state._pre_unshard_stream)
+    state._needs_pre_forward_unshard[handle] = False
     state._device_handle.current_stream().wait_stream(state._unshard_stream)
-    _prefetch_handles(state, handles_key, _PrefetchMode.FORWARD)
+    _prefetch_handle(state, handle, _PrefetchMode.FORWARD)
 
 
 @no_type_check
 def _post_forward(
     state: _FSDPState,
-    handles: List[FlatParamHandle],
+    handle: FlatParamHandle,
```

Comment:
```suggestion
    handle: Optional[FlatParamHandle],
```
(this one actually has some semantic importance since we should still run the post-forward logic, namely registering pre-backward hooks, even if the handle is `None` -- this matters for the root FSDP specifically right now)